Repository/Resource/PhpcrResource.php

```diff
@@ -0,0 +1,174 @@
+<?php
+
+namespace Symfony\Cmf\Component\Resource\Repository\Resource;
+
+use Puli\Repository\Api\Resource\Resource;
+use Puli\Repository\Api\ResourceRepository;
+use PHPCR\NodeInterface;
+use Puli\Repository\Resource\Collection\ArrayResourceCollection;
+use Symfony\Cmf\Component\Resource\Repository\Resource\Metadata\PhpcrMetadata;
+
+/**
+ * Resource representing a PHPCR node
+ *
+ * @author Daniel Leech <daniel@dantleech.com>
+ */
+class PhpcrResource implements Resource
+{
+    private $node;
+    private $path;
+    private $repoPath;
+
+    /**
+     * @param string $path
+     * @param NodeInterface $node
+     */
+    public function __construct($path, NodeInterface $node)
+    {
+        $this->node = $node;
+        $this->repoPath = $path;
+        $this->path = $path;
+    }
+
+    /**
+     * Return the PHPCR node which this resource
+     * represents.
+     *
+     * @return NodeInterface
+     */
+    public function getNode()
+    {
+        return $this->node;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function getPath()
+    {
+        return $this->path;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function getName()
+    {
+        return $this->node->getName();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function getChild($relPath)
+    {
+        $child = $this->node->getNode($relPath);
+        return new self($this->path . '/' . $relPath, $child);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function hasChild($relPath)
+    {
+        return $this->node->hasNode($relPath);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function hasChildren()
+    {
+        return $this->node->hasNodes();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function listChildren()
+    {
+        $collection = new ArrayResourceCollection();
+        foreach ($this->node->getNodes() as $node) {
+            $collection->add(new self($this->path . '/' . $node->getName(), $node));
+        }
+
+        return $collection;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function getMetadata()
+    {
+        return new PhpcrMetadata($this->node);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function getRepository()
+    {
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function getRepositoryPath()
+    {
+        return $this->repoPath;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function attachTo(ResourceRepository $repo, $path = null)
```

Comment:
> a PHPCR resource is dynamic

Sure, you would either need to invalidate the cache when the resource changes or rebuild it periodically.

> what are we caching?

That depends on the repository you are adding the resource to. In the case of the FilesystemRepository, the resource is simply dumped to the filesystem and returned again as a File/DirectoryResource. The metadata will be lost, but in many cases accessing the body only will be sufficient.

You could also implement another type of repository that uses `serialize()` and stores the resource metadata - for example - in a key-value store. There surely are plenty of other use cases and possible cache implementations, depending on what needs to be cached, how often the cache needs to be accessed/refreshed, how fast it needs to be etc.

The point I'm trying to convey is: I want to give the users of Puli the chance to build custom repository implementations that are highly specialized to the needs of their applications. They can then take a generic repository (which may be less efficient, less effective, ...) and copy the resources to their own highly specialized one with very little effort.